### PR TITLE
Reorder dkulib

### PR DESCRIPTION
**Pull request by Muennighoff**
See discussion [here](https://github.com/dataiku/dss-plugin-nlp-amazon-translation/pull/2#discussion_r675523464)

### CODE REVIEW

File: tests/parallelizer/test_parallelizer.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 import pytest
 import pandas as pd
 
-from dkulib.parallelizer.parallelizer import DataFrameParallelizer, ErrorHandling
+from parallelizer.parallelizer import DataFrameParallelizer, ErrorHandling
```
Comment (alexcombessie) [inline tests/parallelizer/test_parallelizer.py:15]: Could we rename the parallelizer module to api_utils? That would better explain the context, and avoid the inelegant parallelizer.parallelizer

Comment (Muennighoff) [reply tests/parallelizer/test_parallelizer.py]: Renamed👍

Comment (Muennighoff) [reply tests/parallelizer/test_parallelizer.py]: Should we do the same for dku_config?

Comment (HenriChabert) [reply tests/parallelizer/test_parallelizer.py]: For dku_config I'd rather not, let's keep it that way

**Comment by alexcombessie** [comment]
Need to fix the tests, and then we're good!

**Comment by Muennighoff** [comment]
> Need to fix the tests, and then we're good!

Yep the problem with the tests is that since now, the dkulib folder is gone `from ..io_utils.plugin_io_utils import generate_unique` is considered as `ValueError: attempted relative import beyond top-level package` - I think because [Python does not consider the current working dir as a package](https://stackoverflow.com/questions/30669474/beyond-top-level-package-error-in-relative-import)

`from io_utils.plugin_io_utils import generate_unique` would work, but won't work once used as a submodule

**Comment by Muennighoff** [comment]
One other option may be to add the dkulib git submodule to the sys path in the repositories its used - Then the absolute import statements should work again, also see [here](https://stackoverflow.com/questions/29746958/how-to-import-python-file-from-git-submodule)

**Comment by alexcombessie** [comment]
> One other option may be to add the dkulib git submodule to the sys path in the repositories its used - Then the absolute import statements should work again, also see [here](https://stackoverflow.com/questions/29746958/how-to-import-python-file-from-git-submodule)

Not sure having the module fiddle with the path is a good idea. I would rather set that explicitly in the makefile, ex: https://github.com/dataiku/dss-plugin-dkulib/blob/fc408033f9531a7823e466c9246ae874e85c1004/Makefile#L15

**Comment by Muennighoff** [comment]
After discussion with @mhham, we confirmed that the problem is that the tests are launched from the PWD, but the relative imports cross the PWD, which is not allowed, as python does not consider the PWD a package.

We brainstormed a lot, so for the record here are our ideas & our conclusion:
- In the Makefile create a subdirectory and move all folders with python logic in there, then launch the tests from the parent dir  ---- Too hacky
- Swap the relative paths such as in parallelizer `from ..io_utils` to absolute paths, such as `from io_utils` & then add the submodule to the sys path when it is used ---- We don't want to meddle with the sys path in the plugins as already the python-lib gets added to it
- Duplicate the necessary utils, e.g. duplicate the io_utils within parallelizer and then import it from downstream in the parallelizer hence not crossing the PWD with a relative path ---- Will require updating multiple duplicated files
- Add an __init__.py to the PWD importing all functions ---- Will also need all requirements of all functions, hence requiring unnecessary packages when used as a submodule

- **Just stick with a dkulib folder with all functionality of this repo & longer import statement in the plugins where this repo is used as a submodule like [here](https://github.com/dataiku/dss-plugin-nlp-amazon-translation/pull/2#discussion_r673819030) ---- This is the solution @mhham and me found to be best for now**

If anybody comes up with a better solution, definitely let us know~

**Comment by alexcombessie** [comment]
> After discussion with @mhham, we confirmed that the problem is that the tests are launched from the PWD, but the relative imports cross the PWD, which is not allowed, as python does not consider the PWD a package.
> 
> We brainstormed a lot, so for the record here are our ideas & our conclusion:
> 
> * In the Makefile create a subdirectory and move all folders with python logic in there, then launch the tests from the parent dir  ---- Too hacky
> * Swap the relative paths such as in parallelizer `from ..io_utils` to absolute paths, such as `from io_utils` & then add the submodule to the sys path when it is used ---- We don't want to meddle with the sys path in the plugins as already the python-lib gets added to it
> * Duplicate the necessary utils, e.g. duplicate the io_utils within parallelizer and then import it from downstream in the parallelizer hence not crossing the PWD with a relative path ---- Will require updating multiple duplicated files
> * Add an **init**.py to the PWD importing all functions ---- Will also need all requirements of all functions, hence requiring unnecessary packages when used as a submodule
> * **Just stick with a dkulib folder with all functionality of this repo & longer import statement in the plugins where this repo is used as a submodule like [here](https://github.com/dataiku/dss-plugin-nlp-amazon-translation/pull/2#discussion_r673819030) ---- This is the solution @mhham and me found to be best for now**
> 
> If anybody comes up with a better solution, definitely let us know~

I am good with this conclusion, thanks for checking the various possibilities. Perhaps, to avoid the ugly `dkulib.dkulib.truc` we could do same renaming and call it `dkulib.python.truc` ? It actually makes practical sense, since dkulib could also contain Java or R code used by plugins.

**Comment by mhham** [comment]
> I am good with this conclusion, thanks for checking the various possibilities. Perhaps, to avoid the ugly `dkulib.dkulib.truc` we could do same renaming and call it `dkulib.python.truc` ? It actually makes practical sense, since dkulib could also contain Java or R code used by plugins.

Since the "root" dkulib is already a python package (with an __init__.py) I find this a bit misleading. However I agree that it would be more readable to avoid `dkulib.dkulib.blah`. Suggestions:

- `dkulib.code.blah`
- `dkulib.src.blah`
- `dkulib.lib.blah`
- `dkulib.core.blah`
- `dkulib.main.blah`

We could also think about splitting dku specific libs from generic ones likes this:
- `dkulib.dku.blah`
- `dkulib.generic.blah`

**Comment by alexcombessie** [comment]
I like `dkulib.core.blah` or `dkulib.main.blah`

**Comment by mhham** [comment]
Since #24 was merged, I think we can safely close this one

**Pull request closed without merging**